File: src/main/java/org/jukeboxmc/form/CustomForm.java
package org.jukeboxmc.form;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.jukeboxmc.form.element.*;

import java.util.ArrayList;
import java.util.List;

/**
 * @author GoMint
 * @version 1.0
 */
public class CustomForm extends Form<FormResponse> {

    private final List<Element> elements = new ArrayList<>();

    public CustomForm( String title ) {
        super( title );
    }

    public Dropdown createDropdown( String id, String text ) {
        Dropdown dropdown = new Dropdown( this, id, text );
        this.elements.add( dropdown );
        this.dirty = true;
        return dropdown;
    }

    public CustomForm addInputField( String id, String text, String placeHolder, String defaultValue ) {
        Input input = new Input( id, text, placeHolder, defaultValue );
        this.elements.add( input );
        this.dirty = true;
        return this;
    }

    public CustomForm addLabel( String text ) {
        Label label = new Label( "", text );
        this.elements.add( label );
        this.dirty = true;
        return this;
    }

    public CustomForm addSlider( String id, String text, float min, float max, float step, float defaultValue ) {
        Slider slider = new Slider( id, text, min, max, step, defaultValue );
        this.elements.add( slider );
        this.dirty = true;
        return this;
    }

    public StepSlider createStepSlider( String id, String text ) {
        StepSlider stepSlider = new StepSlider( this, id, text );
        this.elements.add( stepSlider );
        this.dirty = true;
        return stepSlider;
    }

    public CustomForm addToggle( String id, String text, boolean value ) {
        Toggle toggle = new Toggle( id, text, value );
        this.elements.add( toggle );
        this.dirty = true;
        return this;
    }

    @Override
    public String getFormType() {
        return "custom_form";
    }

    @Override
    public JSONObject toJSON() {
        JSONObject obj = super.toJSON();

        JSONArray content = (JSONArray) obj.get( "content" );
        for ( Element element : this.elements ) {
            content.add( element.toJSON() );
        }

        return obj;
    }

    @Override
    public FormResponse parseResponse( String json ) {
        // Response is an array with values
        try {
            FormResponse response = new FormResponse();
            JSONArray answers = (JSONArray) new JSONParser().parse( json.trim() );
            for ( int i = 0; i < answers.size(); i++ ) {
                Element element = this.elements.get( i );
                response.addAnswer( element.getId(), element.getAnswer( answers.get( i ) ) );
            }

            return response;
        } catch ( Exception e ) {
            return null;
        }
    }

}

File: JukeboxMC-API/src/main/kotlin/org/jukeboxmc/api/form/CustomForm.kt
package org.jukeboxmc.api.form

import com.google.gson.Gson
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import org.jukeboxmc.api.form.element.*

/**
 * @author Kaooot
 * @version 1.0
 */
class CustomForm(title: String) : Form<FormResponse>(title) {

    private val elements = arrayListOf<Element>()

    override fun getFormType(): String = "custom_form"

    override fun parseResponse(json: String): FormResponse {
        val response = FormResponse()
        val answers = Gson().fromJson(json, JsonArray::class.java)
        for (i in 0..answers.size()) {
            val element = this.elements[i]
            response.addAnswer(element.getId(), element.getAnswer(answers[i]))
        }
        return response
    }

    override fun toJson(): JsonObject {
        val jsonObject = super.toJson()
        val content = jsonObject.getAsJsonArray("content")
        for (element in this.elements) {
            content.add(element.toJSON())
        }
        return jsonObject
    }

    fun createDropdown(id: String, text: String): Dropdown {
        val dropDown = Dropdown(this, id, text)
        this.elements.add(dropDown)
        this.setDirty()
        return dropDown
    }

    fun addInputField(id: String, text: String, placeHolder: String, defaultValue: String): CustomForm {
        val inputField = Input(id, text, placeHolder, defaultValue)
        this.elements.add(inputField)
        this.setDirty()
        return this
    }

    fun addLabel(text: String): CustomForm {
        val label = Label("", text)
        this.elements.add(label)
        this.setDirty()
        return this
    }

    fun addSlider(id: String, text: String, min: Float, max: Float, step: Float, defaultValue: Float): CustomForm {
        val slider = Slider(id, text, min, max, step, defaultValue)
        this.elements.add(slider)
        this.setDirty()
        return this
    }

    fun createStepSlider(id: String, text: String): StepSlider {
        val stepSlider = StepSlider(this, id, text)
        this.elements.add(stepSlider)
        this.setDirty()
        return stepSlider
    }

    fun addToggle(id: String, text: String, value: Boolean): CustomForm {
        val toggle = Toggle(id, text, value)
        this.elements.add(toggle)
        this.setDirty()
        return this
    }
}